src/envoy/auth/http_filter.cc

```diff
@@ -13,36 +13,100 @@
  * limitations under the License.
  */
 
-#include <string>
-
 #include "http_filter.h"
+#include "config.h"
+#include "jwt.h"
 
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+#include "envoy/http/async_client.h"
 #include "server/config/network/http_connection_manager.h"
 
+#include "rapidjson/stringbuffer.h"
+#include "rapidjson/writer.h"
+
+#include <string>
+
 namespace Envoy {
 namespace Http {
 
-/*
- * TODO: receive issuer's info & pubkey
- */
-JwtVerificationFilter::JwtVerificationFilter() {}
+namespace {
+
+std::string JsonToString(rapidjson::Document* d) {
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  d->Accept(writer);
+  return buffer.GetString();
+}
+
+}  // namespace
+
+const LowerCaseString& JwtVerificationFilter::AuthorizedHeaderKey() {
+  static LowerCaseString* key = new LowerCaseString("Istio-Auth-UserInfo");
+  return *key;
+}
+
+JwtVerificationFilter::JwtVerificationFilter(
+    std::shared_ptr<Auth::JwtAuthConfig> config)
+    : config_(config) {}
 
 JwtVerificationFilter::~JwtVerificationFilter() {}
 
 void JwtVerificationFilter::onDestroy() {}
 
-FilterHeadersStatus JwtVerificationFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus JwtVerificationFilter::decodeHeaders(HeaderMap& headers,
+                                                         bool) {
+  state_ = Calling;
+  stopped_ = false;
+
   /*
-   * TODO: verify the JWT in the auth header
+   * TODO: update cached public key regularly
    */
-  return FilterHeadersStatus::Continue;
+
+  // list up issuers whose public key should be fetched
+  for (const auto& iss : config_->issuers_) {
+    if (!iss->failed_ && !iss->loaded_) {
+      calling_issuers_[iss->name_] = iss;
+    }
+  }
+  // send HTTP requests to fetch public keys
+  if (!calling_issuers_.empty()) {
```

Comment:
If the first client request starts an async call to fetch public key,  before it is responded,  what do we do for the second client request?

Reply:
It would be done in following PRs for regular updates of public key